In [1]:
import warnings
warnings.filterwarnings('ignore')

# 데이터 읽기를 위한 라이브러리
import numpy as np
np.random.seed(0)
import pandas as pd
import gc, os, time
import scipy as sp
from pandas import DataFrame, Series
from datetime import datetime, date, timedelta
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn import decomposition

# 탐색적 데이터 분석을 위한 라이브러리
import seaborn as sns
import matplotlib.pyplot as plt
from scipy.stats import skew, norm, probplot, boxcox

# 모델링을 위한 라이브러리
from sklearn.linear_model import Ridge
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.feature_selection import mutual_info_regression, VarianceThreshold

import pickle

pd.set_option('display.max_columns',999)

from tqdm import tqdm
from sklearn.model_selection import TimeSeriesSplit

def RMSE(y, y_pred):
    rmse = mean_squared_error(y, y_pred) ** 0.5
    return rmse

In [2]:
# 학습용 데이터 
train_sensor = pd.read_csv('train_sensor.csv')
train_quality = pd.read_csv('train_quality.csv')
# 평가용 데이터 
predict_sensor = pd.read_csv('predict_sensor.csv')

In [3]:
def make_dataset(X, y=None):
    
    # -----------------------------------
    # train_sensor (X 인자)
    # -----------------------------------
    ''' column을 param_alias 로만 pivot table 만들기. '''
    df_X = X.copy()
    df_X = df_X.sort_values(by='end_time',ascending=True)
    df_X['step_id'] = df_X['step_id'].apply(lambda x: str(x).zfill(2))
    # step_id 와 param_alias 를 결합한 임시 컬럼 step_param 을 생성합니다. ex. 17_EPD_para4
    df_X['step_param'] = df_X[['step_id', 'param_alias']].apply(lambda x: '_'.join(x), axis=1)
    df_X_tmp = df_X.pivot_table(index = ['module_name','key_val'], columns = 'step_param', values='mean_val', aggfunc='sum')
    # 데이터 통합을 위해 인덱스를 key_val 로 재설정합니다. 
    df_X_tmp = df_X_tmp.reset_index(level=[0, 1])
    df_X_tmp.set_index('key_val', inplace=True)

    # -----------------------------------
    # 시간 데이터 
    # -----------------------------------
    ''' step별 end_time을 column으로 pivot table 만들기 '''
    df_X['end_time_tmp'] = df_X.apply(lambda x: x['step_id'] + '_end_time', axis=1)
    df_X['end_time'] = pd.to_datetime(df_X['end_time'])
    # end_time 은 센서 데이터가 각 para 별로 서버에 도달한 시간으로 스텝 내 오차가 발생할 수 있습니다. 동일 스텝 구간내 공정 완료 시간이 다른 경우, min 함수를 사용하여 최초 수집된 time을 가져옵니다.
    df_time_tmp = df_X.pivot_table(index = ['key_val'], columns = 'end_time_tmp', values='end_time', aggfunc=lambda x : min(x.unique()))
    df_time_tmp = df_time_tmp.reset_index()
    df_time_tmp.set_index('key_val', inplace=True)

    # -----------------------------------
    # train_quality (y 인자)
    # -----------------------------------

    if y is None : # 평가용 데이터 
        col_target = []
        col_idx = ['module_name', 'key_val']
        df_complete = pd.concat([df_X_tmp, df_time_tmp], axis=1).reset_index()
        df_complete.rename(columns={'index':'key_val'},inplace=True)
    else : # 학습용 데이터 
        df_y = y.copy()
        df_y.set_index('key_val', inplace=True)
        col_target = ['y']
        col_idx = ['module_name', 'key_val', 'end_dt_tm']
        # 센서 데이터, 시간데이터, 품질지표에 대하여 인덱스(key_val)기준으로 데이터프레임을 통합합니다.
        df_complete = pd.concat([df_X_tmp, df_time_tmp, df_y], axis=1).reset_index()
        # 컬럼 이름을 변경합니다.  
        df_complete.rename(columns={'msure_val':'y'}, inplace=True)
        df_complete.rename(columns={'index':'key_val'},inplace=True)


    # 컬럼 순서를 정렬합니다. 
    col_feats = df_X['step_param'].unique().tolist()
    col_feats.sort()
    col_time = [s for s in df_complete.columns.tolist() if "_end_time" in s]
    col_all = col_idx + col_target + col_feats + col_time
    df_complete = df_complete[col_all]
    # 처음 step이 시작된 시점을 기준으로 다시 정렬(APC value를 먹고 들어가는 값을 기준으로 정렬하고 싶었음.)
    df_complete = df_complete.set_index(['module_name','key_val','04_end_time']).sort_index(level=[0,2,1],ascending=True).reset_index()
    df_complete = df_complete[col_all]
    
    # 컬럼을 소문자로 변경합니다. 
    df_complete.columns = df_complete.columns.str.lower()

    return df_complete

# 학습용 데이터 
train = make_dataset(train_sensor, train_quality)
# 평가용 데이터 
predict = make_dataset(predict_sensor)

In [4]:
print(len(train.columns))
print(len(predict.columns))

676
674


In [5]:
# 전체 및 개별 공정 소요시간 변수를 생성하는 함수입니다.
def gen_duration_feats(df, lst_stepsgap):
    
    # 전체 공정 소요시간(초) 변수를 생성합니다. 
    df['gen_tmdiff'] = (df['20_end_time'] - df['04_end_time']).dt.total_seconds()
    
    # 개별 스텝간 공정 소요시간(초) 변수를 생성합니다. 
    # ex. gen_tmdiff_0406 : 04 스텝 공정 완료 시간과 06 스텝 공정 완료 시간의 차이 
    
    for stepgap in lst_stepsgap:
        df[f'gen_tmdiff_{stepgap}'] = (df[f'{stepgap[2:]}_end_time'] - df[f'{stepgap[:2]}_end_time']).dt.total_seconds()

    return df

# 4. 데이터 전처리

In [6]:
# 전처리를 위한 학습용 데이터와 평가용 데이터를 복사합니다.
df_train = train.copy()
df_predict = predict.copy()
del train

In [7]:
# -----------------------------------
# 3 장 EDA 분석에 필요한 변수를 선언합니다.
# -----------------------------------

# 센서 컬럼과 날짜 컬럼을 정의합니다. 
col_sensor = df_train.iloc[:, 4:-7].columns.tolist() 
col_time = df_train.filter(regex='end').columns.tolist() 

assert len(col_sensor) == 665
assert len(col_time) == 8 

# 3.4절 공정 소요시간 분석에 필요한 변수를 정의합니다. 
lst_steps = ['04','06','12','13','17','18', '20']
lst_stepsgap = ['0406','0612','1213','1317','1718','1820']

''' step별로 fdc para명 따로 수집 '''
lst_sensors = []
for step in lst_steps:
    _ = [col for col in col_sensor if col[:2] == step]
    lst_sensors.append(_)

sensors_nm = list(map(lambda x: x[3:], lst_sensors[0]))

# 시간과 관련한 분석을 진행하기 위하여 날짜형으로 변환합니다. 
df_train[col_time] = df_train[col_time].apply(pd.to_datetime)

In [8]:
for_col_filter = []
for step_para in lst_sensors:
    for para in step_para:
        para = para.split('_')[0]+'_'+para.split('_')[1]
        for_col_filter.append(para)
for_col_filter = sorted(list(set(for_col_filter)))

In [9]:
# 전체 및 개별 공정 소요시간 7개의 변수를 생성합니다(3.4절)
lst_stepsgap = ['0406','0612','1213','1317','1718','1820']
df_train = gen_duration_feats(df_train, lst_stepsgap)
df_predict = gen_duration_feats(df_predict, lst_stepsgap)
df_train.filter(regex='tmdiff').head(2)

gen_tmdiff  gen_tmdiff_0406  gen_tmdiff_0612  gen_tmdiff_1213  \
0      1912.0            146.0            846.0             16.0   
1      1911.0            145.0            847.0             16.0   

   gen_tmdiff_1317  gen_tmdiff_1718  gen_tmdiff_1820  
0            477.0             16.0            411.0  
1            476.0             16.0            411.0

# Cyclic Transformation 적용

In [10]:
''' Cyclic Transformation 적용 '''
def cyclic_transformation(df, cols):
    for col in cols:
        step = col[:2]
        df[col] = pd.to_datetime(df[col])
        df[step+'_'+'hour'] = df[col].dt.hour
        df[step+'_'+'month'] = df[col].dt.month
        df[step+'_'+'day'] = df[col].dt.day
        df[step+'_'+'weekday'] = df[col].dt.weekday
        
        ## cyclic transformation on hour
        df[step+'_'+'hour_sin'] = np.sin(2 * np.pi * df[step+'_'+'hour']/23.0)
        df[step+'_'+'hour_cos'] = np.cos(2 * np.pi * df[step+'_'+'hour']/23.0)
        ## cyclic transformation on date 
        df[step+'_'+'date_sin'] = -np.sin(2 * np.pi * (df[step+'_'+'month']+df[step+'_'+'day']/31)/12)
        df[step+'_'+'date_cos'] = -np.cos(2 * np.pi * (df[step+'_'+'month']+df[step+'_'+'day']/31)/12)
        ## cyclic transformation on month
        df[step+'_'+'month_sin'] = -np.sin(2 * np.pi * df[step+'_'+'month']/12.0)
        df[step+'_'+'month_cos'] = -np.cos(2 * np.pi * df[step+'_'+'month']/12.0)
        ## cyclic transformation on weekday
        df[step+'_'+'weekday_sin'] = -np.sin(2 * np.pi * (df[step+'_'+'weekday']+1)/7.0)
        df[step+'_'+'weekday_cos'] = -np.cos(2 * np.pi * (df[step+'_'+'weekday']+1)/7.0)
        
        df.drop(step+'_'+'month',axis=1,inplace=True)
        df.drop(step+'_'+'month_sin',axis=1,inplace=True)
        df.drop(step+'_'+'month_cos',axis=1,inplace=True)

In [11]:
endtime_col = df_train.filter(regex='end_time$').columns.tolist()
cyclic_transformation(df_train, endtime_col)
cyclic_transformation(df_predict, endtime_col)

In [12]:
df_train.head(1)

module_name  key_val           end_dt_tm          y  04_efem_para2  \
0    EQ10_PM1  LOT5_21 2021-10-03 07:10:22  1260.0892       1631.273   

   04_efem_para25  04_efem_para78  04_epd_para4  04_epd_para40  04_epd_para63  \
0        1639.727         5.06654           0.0            0.0            0.0   

   04_epd_para80  04_esc_para84  04_esc_para94  04_fr_para28  04_fr_para35  \
0            0.0         2999.0            0.0       -2999.0           0.0   

   04_fr_para61  04_fr_para69  04_gas_para10  04_gas_para13  04_gas_para15  \
0           0.0        2999.0            0.0            0.0           40.0   

   04_gas_para19  04_gas_para21  04_gas_para26  04_gas_para27  04_gas_para33  \
0            0.0        2.41871       30.04839            0.0            0.0   

   04_gas_para36  04_gas_para39  04_gas_para46  04_gas_para48  04_gas_para50  \
0       2.477097            0.0            0.0            0.0            0.0   

   04_gas_para51  04_gas_para52  04_gas_para59  04_gas_para6  04_gas_para70  \
0           11.6       24.95807        0.09697      2.834516            0.0   

   04_gas_para71  04_gas_para73  04_gas_para74  04_gas_para85  04_he_para1  \
0            0.0       44.90645           50.1       70.26363         15.0   

   04_he_para22  04_he_para88  04_he_para95  04_hv_para3  04_hv_para45  \
0      0.906452      0.726923          30.0         0.09      299.9667   

   04_hv_para47  04_hv_para56  04_position_para72  04_power_para14  \
0         150.0      0.241481            7.033333              0.0   

   04_power_para49  04_power_para57  04_power_para68  04_power_para76  \
0              0.0           1000.0              0.0         1062.933   

   04_power_para82  04_pressure_para91  04_temp_para11  04_temp_para12  \
0              0.0            45.02667        0.598182            25.1   

   04_temp_para17  04_temp_para18  04_temp_para20  04_temp_para23  \
0       -0.021212         15.0697        26.60606        150.1848   

   04_temp_para24  04_temp_para32  04_temp_para38  04_temp_para53  \
0            20.0        19.96061       -9.842424        149.8545   

   04_temp_para54  04_temp_para55  04_temp_para58  04_temp_para60  \
0        47.52941        22.94848            24.5        44.51515   

   04_temp_para65  04_temp_para66  04_temp_para79  04_temp_para86  \
0        21.45152            90.0        0.579091        22.04243   

   04_temp_para87  04_temp_para92  04_temp_para93  04_time_para16  \
0        149.7182        35.01515        149.8545          2460.4   

   04_time_para29  04_time_para30  04_time_para34  04_time_para37  \
0           723.6          2460.4          2460.4        132.3333   

   04_time_para41  04_time_para42  04_time_para43  04_time_para44  \
0        132.3333          2460.4          2460.4          2460.4   

   04_time_para5  04_time_para62  04_time_para64  04_time_para67  \
0            0.0          2460.4        132.3333          2460.4   

   04_time_para7  04_time_para75  04_time_para77  04_time_para8  \
0       132.3333          2460.4          2460.4       132.3333   

   04_time_para81  04_time_para83  04_time_para89  04_time_para9  \
0          2460.4        132.3333          2460.4       132.3333   

   04_time_para90  04_tmp_para31  06_efem_para2  06_efem_para25  \
0          2460.4       2.351515       1631.445        1640.059   

   06_efem_para78  06_epd_para4  06_epd_para40  06_epd_para63  06_epd_para80  \
0        5.074046           0.0      -1.201783            0.0            0.0   

   06_esc_para84  06_esc_para94  06_fr_para28  06_fr_para35  06_fr_para61  \
0         2999.0            0.0       -2999.0           0.0           0.0   

   06_fr_para69  06_gas_para10  06_gas_para13  06_gas_para15  06_gas_para19  \
0        2999.0            0.0            0.0            0.0            0.0   

   06_gas_para21  06_gas_para26  06_gas_para27  06_gas_para33  06_gas_para36  \
0       3.481353       32.89549            0.0            0.0       3.674587   

   

In [13]:
df_predict.head(1)

module_name    key_val  04_efem_para2  04_efem_para25  04_efem_para78  \
0    EQ10_PM1  LOT380_17      1626.8125       1635.9062        5.141955   

   04_epd_para4  04_epd_para40  04_epd_para63  04_epd_para80  04_esc_para84  \
0           0.0            0.0            0.0            0.0         2999.0   

   04_esc_para94  04_fr_para28  04_fr_para35  04_fr_para61  04_fr_para69  \
0            0.0       -2999.0           0.0           0.0        2999.0   

   04_gas_para10  04_gas_para13  04_gas_para15  04_gas_para19  04_gas_para21  \
0            0.0            0.0           40.0            0.0       2.296333   

   04_gas_para26  04_gas_para27  04_gas_para33  04_gas_para36  04_gas_para39  \
0      30.066668            0.0            0.0          2.341            0.0   

   04_gas_para46  04_gas_para48  04_gas_para50  04_gas_para51  04_gas_para52  \
0            0.0            0.0            0.0            9.3      24.963333   

   04_gas_para59  04_gas_para6  04_gas_para70  04_gas_para71  04_gas_para73  \
0            0.1      2.659333            0.0            0.0      44.883335   

   04_gas_para74  04_gas_para85  04_he_para1  04_he_para22  04_he_para88  \
0           50.1      70.296875         15.0           0.8         0.824   

   04_he_para95  04_hv_para3  04_hv_para45  04_hv_para47  04_hv_para56  \
0          30.0          0.0           0.0           0.0           0.0   

   04_position_para72  04_power_para14  04_power_para49  04_power_para57  \
0            7.096552              0.0              0.0           1000.0   

   04_power_para68  04_power_para76  04_power_para82  04_pressure_para91  \
0              0.0        1078.7931              0.0            44.99655   

   04_temp_para11  04_temp_para12  04_temp_para17  04_temp_para18  \
0        0.600625            25.1             0.0       15.084375   

   04_temp_para20  04_temp_para23  04_temp_para24  04_temp_para32  \
0        28.71875       150.16562            20.0       19.971874   

   04_temp_para38  04_temp_para53  04_temp_para54  04_temp_para55  \
0       -9.809375       149.87813       44.969696       22.946875   

   04_temp_para58  04_temp_para60  04_temp_para65  04_temp_para66  \
0            24.5        43.78125       21.421875            90.0   

   04_temp_para79  04_temp_para86  04_temp_para87  04_temp_para92  \
0        0.580313        22.05625        149.8625       35.021873   

   04_temp_para93  04_time_para16  04_time_para29  04_time_para30  \
0       149.87813          2965.1          1228.3          2965.1   

   04_time_para34  04_time_para37  04_time_para41  04_time_para42  \
0            37.8       637.01666            37.8          2965.1   

   04_time_para43  04_time_para44  04_time_para5  04_time_para62  \
0          2965.1          2965.1            0.0          2965.1   

   04_time_para64  04_time_para67  04_time_para7  04_time_para75  \
0            37.8          2965.1      637.01666          2965.1   

   04_time_para77  04_time_para8  04_time_para81  04_time_para83  \
0          2965.1           37.8          2965.1       637.01666   

   04_time_para89  04_time_para9  04_time_para90  04_tmp_para31  \
0          2965.1           37.8            37.8       2.378125   

   06_efem_para2  06_efem_para25  06_efem_para78  06_epd_para4  06_epd_para40  \
0      1627.4822       1637.7142        5.141538      98.89975      -0.114427   

   06_epd_para63  06_epd_para80  06_esc_para84  06_esc_para94  06_fr_para28  \
0       -0.11261            0.0         2999.0            0.0       -2999.0   

   06_fr_para35  06_fr_para61  06_fr_para69  06_gas_para10  06_gas_para13  \
0           0.0           0.0        2999.0            0.0            0.0   

   06_gas_para15  06_gas_para19  06_gas_para21  06_gas_para26  06_gas_para27  \
0            0.0            0.0       3.314455       32.89909            0.0   

   06_gas_para33  06_gas_para36  06_gas_para39  06_gas_para46  06_gas_para48  \
0            0.0       3.538273            0.0      

# Category 변수 처리

In [14]:
''' CATEGORY 변수 처리 및 NUM FEATURE 정의 '''
module2idx = {}
for i, module in enumerate(df_train['module_name'].unique()):
    module2idx[module] = i
    
# eq2idx = {}
# for i, eq in enumerate(df_train['module_name_eq'].unique()):
#     eq2idx[eq] = i
    
def col2cat(df, col, dict):
    df[col] = df[col].apply(lambda x: dict[x])
    df[col] = df[col].astype('category')
    return df[col]

# module_name cat 화
col2cat(df_train, 'module_name', module2idx)
col2cat(df_predict, 'module_name', module2idx)
# eq cat 화
# col2cat(df_train, 'module_name_eq', eq2idx)
# col2cat(df_predict, 'module_name_eq', eq2idx)

0       0
1       0
2       0
3       0
4       0
       ..
257    45
258    46
259    46
260    46
261    46
Name: module_name, Length: 262, dtype: category
Categories (47, int64): [0, 1, 2, 3, ..., 43, 44, 45, 46]

# 전체데이터에 대해서 전처리

In [15]:
df_final = df_train.copy()
df_predict_final = df_predict.copy()

In [16]:
drop_col = []
for para in for_col_filter:
    col = df_final.filter(regex='^'+para).columns.tolist()
    duplicate_deleted_df = df_final[col].T.drop_duplicates(subset=df_final[col].T.columns, keep='first').T
    if len(df_final[col].columns.difference(duplicate_deleted_df.columns))==0:  # 다른게 없으면 무시,
        continue
    else:
        drop_col.extend(df_final[col].columns.difference(duplicate_deleted_df.columns).tolist())
df_final = df_final.drop(drop_col,axis=1)
df_predict_final = df_predict_final.drop(drop_col, axis=1)

In [17]:
print('중복 열 제거 전 columns 개수 : {}'.format(len(df_train.columns)))
print('중복 열 제거 후 columns 개수 : {}'.format(len(df_final.columns)))

중복 열 제거 전 columns 개수 : 746
중복 열 제거 후 columns 개수 : 718


In [18]:
''' feature 정의'''
num_features = list(df_final.columns[df_final.dtypes==float])
num_features.remove('y')
# date_features = list(df_final.columns[df_final.dtypes==np.int64])
col_numerical = num_features
    
''' 분산 0인 col 제거 '''
thresholder = VarianceThreshold(threshold=0)
_ = thresholder.fit_transform(df_final[col_numerical])

 # 분산이 0이면 True 이므로 제거할 컬럼을 추출합니다.  
mask = ~thresholder.get_support()
cols_var_drop = np.asarray(col_numerical)[mask].tolist()
print(f'** {len(cols_var_drop)} Features to Drop by Low Variance')
print(f'{cols_var_drop}')

** 37 Features to Drop by Low Variance
['04_epd_para4', '04_he_para1', '04_power_para14', '04_temp_para24', '04_time_para5', '06_epd_para80', '06_he_para1', '06_hv_para3', '06_power_para14', '06_temp_para24', '06_time_para5', '12_epd_para4', '12_he_para1', '12_he_para95', '12_power_para49', '12_temp_para54', '13_epd_para4', '13_he_para1', '13_he_para95', '13_power_para49', '13_temp_para54', '17_epd_para4', '17_he_para1', '17_he_para95', '17_power_para49', '17_temp_para54', '18_epd_para4', '18_he_para1', '18_he_para95', '18_power_para49', '18_temp_para54', '20_epd_para40', '20_he_para1', '20_he_para95', '20_hv_para47', '20_power_para49', '20_temp_para54']


In [19]:
df_final.drop(cols_var_drop,axis=1,inplace=True)
df_predict_final.drop(cols_var_drop,axis=1,inplace=True)

# 모델링

In [20]:
from catboost import CatBoostRegressor
import xgboost as xgb

import optuna
from optuna.samplers import TPESampler
from sklearn.model_selection import LeaveOneOut, KFold ,cross_val_score
from sklearn.metrics import mean_squared_error

In [21]:
''' feature 재정의'''
without_date_num_features = list(df_final.columns[df_final.dtypes==float])
cat_features = ['module_name']
# date_features = list(df_final.columns[df_final.dtypes==np.int64])
COLS_train = without_date_num_features + cat_features
without_date_num_features.remove('y')
COLS_pred = without_date_num_features + cat_features

In [22]:
def prep_cate_feats(df_tr, df_te, feat_nm):

    df_merge = pd.concat([df_tr, df_te])

    # 컬럼명과 범주형 변수의 레벨명을 이용한 새로운 컬럼명을 자동생성합니다. 
    # ex. module_name_eq -> module_name_eq_EQ01, module_name_eq_EQ02, etc.
    df_merge = pd.get_dummies(df_merge, columns=[feat_nm])

    df_tr = df_merge.iloc[:df_tr.shape[0], :].reset_index(drop=True)
    df_te = df_merge.iloc[df_tr.shape[0]:, :].reset_index(drop=True)

    return df_tr, df_te

# module_name_eq 의 원-핫 인코딩 변수를 생성합니다.
final_one_hot, predict_one_hot = prep_cate_feats(df_final[COLS_train], df_predict_final[COLS_pred], 'module_name')

In [23]:
print('One Hot 전 columns 개수 : {}'.format(len(df_final[COLS_train].columns)))
print('One Hot 후 columns 개수 : {}'.format(len(final_one_hot.columns)))

One Hot 전 columns 개수 : 651
One Hot 후 columns 개수 : 697


In [27]:
one_hot_module_cols = final_one_hot.filter(regex='^module_name').columns.tolist()
X_cols = without_date_num_features + one_hot_module_cols

# 전체 XGB

In [28]:
def objective(trial):
    params_xgb = {
        'booster':trial.suggest_categorical('booster',['gbtree','dart']),
        "reg_lambda": trial.suggest_float("reg_lambda", 0, 1.0),
        'colsample_bytree': trial.suggest_int('colsample_bytree', 0.3,1.0),
        'subsample': trial.suggest_float('subsample', 0.3, 1.0),
        'learning_rate': trial.suggest_loguniform('learning_rate', 0.01, 0.5),
        'n_estimators': trial.suggest_int('n_estimators', 100, 10000),
        'max_depth': trial.suggest_int("max_depth", 4, 12),
        'random_state': trial.suggest_categorical('random_state', [0]),
        'min_child_weight': trial.suggest_int('min_child_weight', 1, 300),
        'tree_method':'gpu_hist',
        'gpu_id':'0'
    }
    
    # 학습 데이터 중 일부를 검증 데이터 셋으로 분할합니다. 
    X_train, X_valid, y_train, y_valid = train_test_split(final_one_hot[X_cols], np.log1p(final_one_hot['y']), test_size=0.15, shuffle=True, random_state=71)

    model = xgb.XGBRegressor(**params_xgb)
    model.fit(
        X_train, y_train,
        eval_set=[(X_valid, y_valid)],
        early_stopping_rounds=35,
        verbose=False
    )

    xgb_pred = model.predict(X_valid)
    rmsle_val = rmse(y_valid, xgb_pred)
    
    return rmsle_val

In [29]:
sampler = TPESampler(seed=42)
study = optuna.create_study(
    study_name="xgb_parameter_opt",
    direction="minimize",
    sampler=sampler,
)
study.optimize(objective, n_trials=30)
print("Best Score:", study.best_value)
print("Best trial:", study.best_trial.params)

[I 2022-07-10 22:51:44,916] A new study created in memory with name: xgb_parameter_opt


KeyboardInterrupt: 

# 전체 CatBoostRegressor

In [46]:
def objective_CAT(trial):
    param = {
      "random_state":42,
      'learning_rate' : trial.suggest_loguniform('learning_rate', 0.01, 0.5),
      'bagging_temperature' :trial.suggest_loguniform('bagging_temperature', 0.01, 100.00),
#       "n_estimators":trial.suggest_int("n_estimators", 100, 10000),
      "max_depth":trial.suggest_int("max_depth", 4, 12),
      'random_strength' :trial.suggest_int('random_strength', 0, 30),
#       "colsample_bylevel":trial.suggest_float("colsample_bylevel", 0.4, 1.0),
      "l2_leaf_reg":trial.suggest_float("l2_leaf_reg",1e-8,3e-5),
      "min_child_samples": trial.suggest_int("min_child_samples", 5, 100),
      "max_bin": trial.suggest_int("max_bin", 200, 400),
      'od_type': trial.suggest_categorical('od_type', ['IncToDec', 'Iter']),
      'boosting_type':trial.suggest_categorical('boosting_type', ['Plain', 'Ordered']),
      'task_type':'GPU',
      'devices':'0:7',
      'iterations':100,
      'rsm':1
  }
    
    X_train, X_valid, y_train, y_valid = train_test_split(df_final[COLS_pred], np.log1p(df_final['y']), test_size=0.15, shuffle=True, random_state=71)
  
    cat = CatBoostRegressor(loss_function='RMSE', eval_metric='RMSE', **param)
    cat.fit(X_train, y_train,
            eval_set=[(X_valid,y_valid)],
            early_stopping_rounds=35, cat_features=cat_features,
            verbose=100)
    cat_pred = cat.predict(X_valid)
    rmsle_val = np.sqrt(mean_squared_error(y_valid, cat_pred))
    
    return rmsle_val

In [ ]:
sampler = TPESampler(seed=42)
study = optuna.create_study(
    study_name="cat_parameter_opt",
    direction="minimize",
    sampler=sampler,
)
study.optimize(objective_CAT, n_trials=20)
print("Best Score:", study.best_value)
print("Best trial:", study.best_trial.params)
# 6.36 -> 9.4 됨.

[I 2022-07-10 23:50:49,308] A new study created in memory with name: cat_parameter_opt


0:	learn: 0.0066316	test: 0.0055544	best: 0.0055544 (0)	total: 17.5ms	remaining: 1.74s


[I 2022-07-10 23:50:50,486] Trial 0 finished with value: 0.005554436407106352 and parameters: {'learning_rate': 0.043284502212938815, 'bagging_temperature': 63.512210106407046, 'max_depth': 10, 'random_strength': 18, 'l2_leaf_reg': 4.688999026868671e-06, 'min_child_samples': 19, 'max_bin': 211, 'od_type': 'IncToDec', 'boosting_type': 'Plain'}. Best is trial 0 with value: 0.005554436407106352.


bestTest = 0.005554434329
bestIteration = 0
Shrink model to first 1 iterations.
0:	learn: 0.0064012	test: 0.0056118	best: 0.0056118 (0)	total: 12.7ms	remaining: 1.26s


[I 2022-07-10 23:50:51,801] Trial 1 finished with value: 0.0053635930358311365 and parameters: {'learning_rate': 0.44447541666908114, 'bagging_temperature': 21.368329072358772, 'max_depth': 5, 'random_strength': 5, 'l2_leaf_reg': 5.51030125050448e-06, 'min_child_samples': 34, 'max_bin': 305, 'od_type': 'IncToDec', 'boosting_type': 'Plain'}. Best is trial 1 with value: 0.0053635930358311365.


bestTest = 0.005363653329
bestIteration = 13
Shrink model to first 14 iterations.
0:	learn: 0.0066000	test: 0.0055273	best: 0.0055273 (0)	total: 27.9ms	remaining: 2.76s


[I 2022-07-10 23:50:54,320] Trial 2 finished with value: 0.005328655676638196 and parameters: {'learning_rate': 0.03135775732257745, 'bagging_temperature': 0.2920433847181412, 'max_depth': 8, 'random_strength': 24, 'l2_leaf_reg': 5.998216726929208e-06, 'min_child_samples': 54, 'max_bin': 319, 'od_type': 'Iter', 'boosting_type': 'Plain'}. Best is trial 2 with value: 0.005328655676638196.


bestTest = 0.005328658887
bestIteration = 31
Shrink model to first 32 iterations.


# LinearModels Modeling

In [ ]:
from sklearn.linear_model import Ridge,ElasticNet,BayesianRidge,Lasso

#### Ridge

In [ ]:
ridges = []
for i, (train, num_f) in enumerate(zip(df_trains, num_features_lst)):
    def objective_ridge(trial):
        param = {
          "random_state":42,
            'alpha':trial.suggest_float("alpha",0.1,10),
            'fit_intercept':trial.suggest_categorical('fit_intercept', [True, False]),
            'normalize':trial.suggest_categorical('normalize', [True, False]),
        }
        X = train[num_f]
        y = np.log1p(train['y'])

        model = Ridge(**param)
        X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.2, shuffle=True, random_state=71)

        model.fit(
            X_train, y_train
            )
        ridge_pred = model.predict(X_valid)
        rmsle_val = np.sqrt(mean_squared_error(y_valid, ridge_pred))

        return rmsle_val
    
    sampler = TPESampler(seed=42)
    study = optuna.create_study(
            study_name="ridge_parameter_opt",
            direction="minimize",
            sampler=sampler,
            )
    study.optimize(objective_ridge, n_trials=10)
    print("Best Score:", study.best_value)
    print("Best trial:", study.best_trial.params)
    
    model = Ridge(**study.best_params)
    model.fit(train[num_f], np.log1p(train['y']))
    print('{} model training is completed'.format(i))
    ridges.append(model)


#### ElasticNet

In [ ]:
ens = []
for i, (train, num_f) in enumerate(zip(df_trains, num_features_lst)):
    def objective_en(trial):
        param = {
          "random_state":42,
            'alpha':trial.suggest_float("alpha",0.1,20),
            'fit_intercept':trial.suggest_categorical('fit_intercept', [True, False]),
            'normalize':trial.suggest_categorical('normalize', [True, False]),
        }
        X = train[num_f]
        y = np.log1p(train['y'])

        model = ElasticNet(**param)
        X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.2, shuffle=True, random_state=71)

        model.fit(
            X_train, y_train
            )
        en_pred = model.predict(X_valid)
        rmsle_val = np.sqrt(mean_squared_error(y_valid, en_pred))

        return rmsle_val
    
    sampler = TPESampler(seed=42)
    study = optuna.create_study(
            study_name="ridge_parameter_opt",
            direction="minimize",
            sampler=sampler,
            )
    study.optimize(objective_en, n_trials=10)
    print("Best Score:", study.best_value)
    print("Best trial:", study.best_trial.params)
    
    model = ElasticNet(**study.best_params)
    model.fit(train[num_f], np.log1p(train['y']))
    print('{} model training is completed'.format(i))
    ens.append(model)


#### BayesianRidge

In [ ]:
brs = []
for i, (train, num_f) in enumerate(zip(df_trains, num_features_lst)):
    def objective_br(trial):
        param = {
            'n_iter':trial.suggest_int("n_iter",10,500),
            'alpha_2':trial.suggest_uniform("alpha_2",-10,10),
            'lambda_2' :trial.suggest_uniform('lambda_2', -10, 10),
            'fit_intercept':trial.suggest_categorical('fit_intercept', [True, False]),
            'normalize':trial.suggest_categorical('normalize', [True, False]),
        }
        X = train[num_f]
        y = np.log1p(train['y'])

        model = BayesianRidge(**param)
        X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.2, shuffle=True, random_state=71)

        model.fit(
            X_train, y_train
            )
        br_pred = model.predict(X_valid)
        rmsle_val = np.sqrt(mean_squared_error(y_valid, br_pred))

        return rmsle_val
    
    sampler = TPESampler(seed=42)
    study = optuna.create_study(
            study_name="BayesianRidge_parameter_opt",
            direction="minimize",
            sampler=sampler,
            )
    study.optimize(objective_br, n_trials=10)
    print("Best Score:", study.best_value)
    print("Best trial:", study.best_trial.params)
    
    model = BayesianRidge(**study.best_params)
    model.fit(train[num_f], np.log1p(train['y']))
    print('{} model training is completed'.format(i))
    brs.append(model)


#### Lasso

In [ ]:
ls = []
for i, (train, num_f) in enumerate(zip(df_trains, num_features_lst)):
    def objective_ls(trial):
        param = {
          "random_state":42,
            'alpha':trial.suggest_float("alpha",0.1,20),
            'fit_intercept':trial.suggest_categorical('fit_intercept', [True, False]),
            'normalize':trial.suggest_categorical('normalize', [True, False]),
            'warm_start':True
        }
        X = train[num_f]
        y = np.log1p(train['y'])

        model = Lasso(**param, warm_start=True)
        X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.2, shuffle=True, random_state=71)

        model.fit(
            X_train, y_train
            )
        ls_pred = model.predict(X_valid)
        rmsle_val = np.sqrt(mean_squared_error(y_valid, ls_pred))

        return rmsle_val
    
    sampler = TPESampler(seed=42)
    study = optuna.create_study(
            study_name="Lasso_parameter_opt",
            direction="minimize",
            sampler=sampler,
            )
    study.optimize(objective_ls, n_trials=10)
    print("Best Score:", study.best_value)
    print("Best trial:", study.best_trial.params)
    
    model = Lasso(**study.best_params, warm_start=True)
    model.fit(train[num_f], np.log1p(train['y']))
    print('{} model training is completed'.format(i))
    ls.append(model)


In [ ]:
models = [cats,ridges,ens,brs,ls]
for i,model_all_chamber in enumerate(models):
    for j, (df,model_by_chamber,num_f) in enumerate(zip(df_predicts, model_all_chamber, num_features_lst)):
        pred_by_chamber = model_by_chamber.predict(df[num_f])
        predict.loc[df.index, 'pred_{}'.format(i)] = pred_by_chamber

In [ ]:
predict.isnull().values.any()

In [ ]:
predict['msure_val'] = np.mean(np.exp(predict.filter(regex='^pred')), axis=1)
df_submission = predict[['key_val','msure_val']]
df_submission.head()

In [ ]:
# 예측 파일을 저장합니다. 
# 제출용 파일 이름은 cds_submission_팀명_차수.csv 형태로 제출합니다.
df_submission.set_index('key_val', inplace=True)
df_submission.to_csv('cds_submission_데이터조무사_25.csv')

# PYCARET

In [30]:
from pycaret.regression import *

In [31]:
for_pycaret = final_one_hot[X_cols+['y']].copy()
for_pycaret.head(1)

04_efem_para2  04_efem_para25  04_efem_para78  04_esc_para84  \
0       1631.273        1639.727         5.06654         2999.0   

   04_esc_para94  04_fr_para28  04_fr_para35  04_fr_para61  04_fr_para69  \
0            0.0       -2999.0           0.0           0.0        2999.0   

   04_gas_para10  04_gas_para13  04_gas_para15  04_gas_para19  04_gas_para21  \
0            0.0            0.0           40.0            0.0        2.41871   

   04_gas_para26  04_gas_para27  04_gas_para33  04_gas_para36  04_gas_para39  \
0       30.04839            0.0            0.0       2.477097            0.0   

   04_gas_para46  04_gas_para48  04_gas_para50  04_gas_para51  04_gas_para52  \
0            0.0            0.0            0.0           11.6       24.95807   

   04_gas_para59  04_gas_para6  04_gas_para70  04_gas_para71  04_gas_para73  \
0        0.09697      2.834516            0.0            0.0       44.90645   

   04_gas_para74  04_gas_para85  04_he_para22  04_he_para88  04_he_para95  \
0           50.1       70.26363      0.906452      0.726923          30.0   

   04_hv_para3  04_hv_para45  04_hv_para47  04_hv_para56  04_position_para72  \
0         0.09      299.9667         150.0      0.241481            7.033333   

   04_power_para57  04_power_para76  04_pressure_para91  04_temp_para11  \
0           1000.0         1062.933            45.02667        0.598182   

   04_temp_para12  04_temp_para17  04_temp_para18  04_temp_para20  \
0            25.1       -0.021212         15.0697        26.60606   

   04_temp_para23  04_temp_para32  04_temp_para38  04_temp_para53  \
0        150.1848        19.96061       -9.842424        149.8545   

   04_temp_para54  04_temp_para55  04_temp_para58  04_temp_para60  \
0        47.52941        22.94848            24.5        44.51515   

   04_temp_para65  04_temp_para66  04_temp_para79  04_temp_para86  \
0        21.45152            90.0        0.579091        22.04243   

   04_temp_para87  04_temp_para92  04_temp_para93  04_time_para16  \
0        149.7182        35.01515        149.8545          2460.4   

   04_time_para29  04_time_para30  04_time_para34  04_time_para37  \
0           723.6          2460.4          2460.4        132.3333   

   04_time_para41  04_time_para42  04_time_para43  04_time_para44  \
0        132.3333          2460.4          2460.4          2460.4   

   04_time_para62  04_time_para64  04_time_para67  04_time_para7  \
0          2460.4        132.3333          2460.4       132.3333   

   04_time_para75  04_time_para77  04_time_para8  04_time_para81  \
0          2460.4          2460.4       132.3333          2460.4   

   04_time_para83  04_time_para89  04_time_para9  04_time_para90  \
0        132.3333          2460.4       132.3333          2460.4   

   04_tmp_para31  06_efem_para2  06_efem_para25  06_efem_para78  06_epd_para4  \
0       2.351515       1631.445        1640.059        5.074046           0.0   

   06_epd_para40  06_epd_para63  06_esc_para84  06_esc_para94  06_fr_para28  \
0      -1.201783            0.0         2999.0            0.0       -2999.0   

   06_fr_para35  06_fr_para61  06_fr_para69  06_gas_para10  06_gas_para13  \
0           0.0           0.0        2999.0            0.0            0.0   

   06_gas_para15  06_gas_para19  06_gas_para21  06_gas_para26  06_gas_para27  \
0            0.0            0.0       3.481353       32.89549            0.0   

   06_gas_para33  06_gas_para36  06_gas_para39  06_gas_para46  06_gas_para48  \
0            0.0       3.674587            0.0            0.0            0.0   

   06_gas_para50  06_gas_para51  06_gas_para52  06_gas_para59  06_gas_para6  \
0            0.0            0.0       34.09699        179.697      3.162481   

   06_gas_para70  06_gas_para71  06_gas_para73  06_gas_para74  06_gas_para85  \
0           15.0       120.9215       33.00075            0.0            0.0   

   06_he_para22  06_he_para88  06_he_para95  06_position_para72  \
0      0.900752      0.704688       

In [32]:
sup = setup(for_pycaret, target = 'y', train_size = 0.8)

In [33]:
comp = compare_models(sort = 'RMSE')

In [34]:
top1 = create_model('ridge', cross_validation = False)
top2 = create_model('en', cross_validation = False)
top3 = create_model('lightgbm', cross_validation = False)
top4 = create_model('catboost', cross_validation = False)
top5 = create_model('lasso', cross_validation = False)

MAE        MSE    RMSE      R2   RMSLE    MAPE
0  5.3457  46.810398  6.8418  0.1828  0.0054  0.0042

In [35]:
tuned_ridge = tune_model(top1, optimize = 'RMSE', n_iter = 10)
tuned_en = tune_model(top2, optimize = 'RMSE', n_iter = 10)
tuned_lgb = tune_model(top3, optimize = 'RMSE', n_iter = 10)
tuned_cat = tune_model(top4, optimize = 'RMSE', n_iter = 10)
tuned_lasso = tune_model(top5, optimize = 'RMSE', n_iter = 10)

### TOP5 Blending

In [36]:
''' top5 blending'''
blender_specific5 = blend_models(estimator_list = [tuned_ridge,tuned_en,tuned_lgb,tuned_cat,tuned_lasso],optimize = 'RMSE')

In [37]:
final_model5 = finalize_model(blender_specific5)
pred = predict_model(final_model5, data=predict_one_hot[X_cols])

In [38]:
predict['msure_val'] = pred['Label']
df_submission = predict[['key_val', 'msure_val']]
df_submission.head()

key_val    msure_val
0  LOT380_17  1261.143394
1  LOT122_18  1260.694982
2  LOT313_18  1260.446343
3  LOT459_12  1260.933767
4  LOT459_18  1263.855695

In [39]:
df_submission.set_index('key_val', inplace=True)
df_submission.to_csv('cds_submission_데이터조무사_26.csv')

### TOP7 Blending

In [40]:
top6 = create_model('br', cross_validation = False)
top7 = create_model('et', cross_validation = False)
# top8 = create_model('rf', cross_validation = False)
# top9 = create_model('ada', cross_validation = False)
# top10 = create_model('gbr', cross_validation = False)

MAE      MSE    RMSE      R2   RMSLE    MAPE
0  5.4299  49.9592  7.0682  0.1278  0.0056  0.0043

In [41]:
tuned_br = tune_model(top6, optimize = 'RMSE', n_iter = 10)
tuned_et = tune_model(top7, optimize = 'RMSE', n_iter = 10)
# tuned_rf = tune_model(top8, optimize = 'RMSE', n_iter = 10)
# tuned_ada = tune_model(top9, optimize = 'RMSE', n_iter = 10)
# tuned_gbr = tune_model(top10, optimize = 'RMSE', n_iter = 10)

In [42]:
''' top7 blending'''
blender_specific7 = blend_models(estimator_list = [tuned_ridge,tuned_en,tuned_lgb,tuned_cat,tuned_lasso,tuned_br,tuned_et],optimize = 'RMSE')

In [43]:
final_model7 = finalize_model(blender_specific7)
pred = predict_model(final_model7, data=predict_one_hot[X_cols])

In [44]:
predict['msure_val'] = pred['Label']
df_submission = predict[['key_val', 'msure_val']]
df_submission.head()

key_val    msure_val
0  LOT380_17  1261.089682
1  LOT122_18  1260.776507
2  LOT313_18  1260.406764
3  LOT459_12  1261.072806
4  LOT459_18  1263.309557

In [45]:
# 예측 파일을 저장합니다. 
# 제출용 파일 이름은 cds_submission_팀명_차수.csv 형태로 제출합니다.
df_submission.set_index('key_val', inplace=True)
df_submission.to_csv('cds_submission_데이터조무사_27.csv')

# Feature Selection

In [ ]:
from probatus.feature_elimination import EarlyStoppingShapRFECV

In [ ]:
# Run feature elimination
shap_elimination = EarlyStoppingShapRFECV(
    clf=model_xgb, step=0.2, cv=10, scoring='neg_mean_squared_error', early_stopping_rounds=15, n_jobs=-1, eval_metric='rmse')
report = shap_elimination.fit_compute(df_final[COLS], np.log1p(df_final['y']))

# Make plots
performance_plot = shap_elimination.plot()

In [ ]:
report

In [ ]:
# Get final feature set
''' num_features는 마지막에 남길 feature 수임. '''
final_features_set = shap_elimination.get_reduced_features_set(num_features=77)

The provided number of features has not been achieved at any stage of the process. You can select one of the following: [402, 322, 258, 207, 166, 133, 107, 86, 69, 56, 45, 36, 29, 24, 20, 16, 13, 11, 9, 8, 7, 6, 5, 4, 3, 2, 1]

In [ ]:
final_features_set

# 예측 결과 제출

In [ ]:
df_predict.head()

In [ ]:
pred_X = pd.get_dummies(df_predict[COLS])

In [ ]:
predict['msure_val'] = np.exp(cat.predict(df_predict_final[COLS]))
df_submission = predict[['key_val', 'msure_val']] 
df_submission.head()

In [ ]:
# 예측값에 결측치가 포함되어 있는지 확인합니다.
df_submission.isnull().sum()

In [ ]:
# 예측값의 갯수가 평가용 데이터의 갯수와 동일한지 확인합니다.
assert len(df_submission) == len(predict)
print(f'No. of Predict DataSet : {len(predict)}\nNo. of Submission DataSet : {len(df_submission)}')

In [ ]:
# 예측 파일을 저장합니다. 
# 제출용 파일 이름은 cds_submission_팀명_차수.csv 형태로 제출합니다.
df_submission.set_index('key_val', inplace=True)
df_submission.to_csv('cds_submission_데이터조무사_12.csv')